datastore/longhorn.go

```diff
@@ -402,21 +415,72 @@ func (s *DataStore) ValidateSetting(name, value string) (err error) {
 	return nil
 }
 
-func (s *DataStore) ValidateV2DataEngine(v2DataEngineEnabled bool) error {
-	if !v2DataEngineEnabled {
-		allV2VolumesDetached, err := s.AreAllV2VolumesDetached()
+func (s *DataStore) ValidateV1DataEngine(dataEngineEnabled bool) error {
+	if !dataEngineEnabled {
+		allVolumesDetached, err := s.AreAllVolumesDetachedByBackendStoreDriver(longhorn.BackendStoreDriverTypeV1)
+		if err != nil {
+			return errors.Wrapf(err, "failed to check volume detachment for %v setting update", types.SettingNameV1DataEngine)
+		}
+		if !allVolumesDetached {
+			return &types.ErrorInvalidState{Reason: fmt.Sprintf("cannot apply %v setting to Longhorn workloads when there are attached volumes", types.SettingNameV1DataEngine)}
+		}
+
+		allDisksRemoved, err := s.AreAllDisksRemovedByDiskType(longhorn.DiskTypeFilesystem)
+		if err != nil {
+			return errors.Wrapf(err, "failed to check block-type disk removal for %v setting update", types.SettingNameV1DataEngine)
+		}
+		if !allDisksRemoved {
+			return &types.ErrorInvalidState{Reason: fmt.Sprintf("cannot apply %v setting to Longhorn workloads when there are filesystem-type disks", types.SettingNameV1DataEngine)}
+		}
+		return nil
+	}
+
+	ims, err := s.ListInstanceManagersRO()
+	if err != nil {
+		return errors.Wrapf(err, "failed to list instance managers for %v setting update", types.SettingNameV2DataEngine)
+	}
+
+	for _, im := range ims {
+		node, err := s.GetKubernetesNodeRO(im.Spec.NodeID)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return errors.Wrapf(err, "failed to get Kubernetes node %v for %v setting update", im.Spec.NodeID, types.SettingNameV2DataEngine)
+			}
+			continue
+		}
+
+		lhNode, err := s.GetNodeRO(im.Spec.NodeID)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get Longhorn node %v for %v setting update", im.Spec.NodeID, types.SettingNameV2DataEngine)
+		}
+
+		if !dataEngineEnabled {
+			for _, disk := range lhNode.Spec.Disks {
+				if disk.Type == longhorn.DiskTypeFilesystem {
+					return fmt.Errorf("failed to disable %v setting because there are filesystem-type disks on node %v", types.SettingNameV2DataEngine, node.Name)
+				}
+			}
+		}
+	}
```

Comment:
Why do we need to check instance manager pods?
Why do users need to remove all longhorn FS disks before disabling the v1 engine?

Don't we just let everything intact instead? All disks and data inside still persist even though the engine has been disabled.

The same questions for v2 engine.

Reply:
Why do we need to check instance manager pods?
> Why do users need to remove all longhorn FS disks before disabling the v1 engine?

The check for v1 engine is not required, but it is necessary for v2 engine because the disk is managed by spdk_tgt.

Reply:
Removed the unnecessary checks. Thank you.